Testnet/DAOContract/DAOContract/DAOContract.cs

```diff
@@ -0,0 +1,246 @@
+    }
+
+    public bool IsWhitelisted(Address address) => State.GetBool($"Whitelisted:{address}");
+    private void SetIsWhitelisted(Address address, bool allowed) => State.SetBool($"Whitelisted:{address}", allowed);
+
+    public ulong GetVotingDeadline(uint proposalId) => State.GetUInt64($"Deadline:{proposalId}");
+    private void SetVotingDeadline(uint proposalId, ulong block) => State.SetUInt64($"Deadline:{proposalId}", block);
+
+    public uint GetYesVotes(uint proposalId) => State.GetUInt32($"YesVotes:{proposalId}");
+    private void SetYesVotes(uint proposalId, uint value) => State.SetUInt32($"YesVotes:{proposalId}", value);
+
+    public uint GetNoVotes(uint proposalId) => State.GetUInt32($"NoVotes:{proposalId}");
+    private void SetNoVotes(uint proposalId, uint value) => State.SetUInt32($"NoVotes:{proposalId}", value);
+
+    public uint GetVote(uint proposalId, Address address) => State.GetUInt32($"Vote:{proposalId}:{address}");
+
+    public Proposal GetProposal(uint index) => State.GetStruct<Proposal>($"Proposals:{index}");
+
+    private void SetProposal(uint index, Proposal proposal) => State.SetStruct($"Proposals:{index}", proposal);
+
+    public DAOContract(ISmartContractState state, uint minQuorum)
+        : base(state)
+    {
+        this.Owner = Message.Sender;
+        this.MinQuorum = minQuorum;
+        this.LastProposalId = 1;
+    }
+
+    public uint CreateProposal(Address recipent, ulong amount, uint votingDuration, string description)
+    {
+        var length = description?.Length ?? 0;
+        Assert(length <= 200,"The description length can be up to 200 characters.");
+
+        var proposal = new Proposal
+        {
+            RequestedAmount = amount,
+            Description = description,
+            Recipient = recipent,
+            Owner = Message.Sender
+        };
+
+        var proposalId = LastProposalId;
+        SetProposal(proposalId, proposal);
+        SetVotingDeadline(proposalId, checked(votingDuration + Block.Number));
+        Log(new ProposalAddedLog
+        {
+            ProposalId = proposalId,
+            Recipent = recipent,
+            Amount = amount,
+            Description = description
+        });
+
+        LastProposalId = proposalId + 1;
+
+        return proposalId;
+    }
+
+    public void Vote(uint proposalId, bool vote)
+    {
+        Assert(IsWhitelisted(Message.Sender), "The caller is not whitelisted.");
+
+        Assert(GetVotingDeadline(proposalId) > Block.Number, "Voting is closed.");
+
+        SetVote(proposalId, Message.Sender, ToVote(vote));
+    }
+
+    private void SetVote(uint proposalId, Address address, Votes vote)
+    {
+        var currentVote = (Votes)GetVote(proposalId, address);
+
+        if (currentVote == vote)
+        {
+            return;
+        }
+
+        Unvote(proposalId, currentVote);
+        State.SetUInt32($"Vote:{proposalId}:{address}", (uint)vote);
```

Comment:
For consistency, maybe a private `SetVote` method.

Reply:
SetVote is actually private setter which has some additional logic

Reply:
ok you can resolve this, I can't 👍 thanks

Reply:
Done

Reply:
mark this comment as resolved is what I meant. I don't have permissions